From 9518d4ad0f3261c35cbbf24222f81562e495f60a Mon Sep 17 00:00:00 2001
From: Neil Mayhew <neil_mayhew@sil.org>
Date: Wed, 28 Jan 2015 15:48:21 -0700
Subject: Upgrade samples/docview to open files on start-up

On most systems, files are given on the command line, and if no files
are given, a new file should be created.

On Mac, start-up actions are not given via the command line but are sent
as Apple Events. Either an Open Files or a New File Apple Event will be
sent. These are handled by wxApp::MacOpenFiles and wxApp::MacNewFile.
However, wxApp::MacNewFile does nothing, and must be overridden in a
derived class to obtain the correct behaviour.
---
 samples/docview/docview.cpp | 19 ++++++++++++++++++-
 samples/docview/docview.h   |  6 ++++++
 2 files changed, 24 insertions(+), 1 deletion(-)

diff --git a/samples/docview/docview.cpp b/samples/docview/docview.cpp
index 7402328..2269f05 100644
--- a/samples/docview/docview.cpp
+++ b/samples/docview/docview.cpp
@@ -108,6 +108,8 @@ void MyApp::OnInitCmdLine(wxCmdLineParser& parser)
                      "run in SDI mode: multiple documents, multiple windows");
     parser.AddSwitch("", CmdLineOption::SINGLE,
                      "run in single document mode");
+
+    parser.AddParam("FILE", wxCMD_LINE_VAL_STRING, wxCMD_LINE_PARAM_MULTIPLE | wxCMD_LINE_PARAM_OPTIONAL);
 }
 
 bool MyApp::OnCmdLineParsed(wxCmdLineParser& parser)
@@ -140,9 +142,18 @@ bool MyApp::OnCmdLineParsed(wxCmdLineParser& parser)
         return false;
     }
 
+    // save params for later use
+    for (size_t i = 0; i != parser.GetParamCount(); ++i)
+        m_initialFiles.push_back(parser.GetParam(i));
+
     return wxApp::OnCmdLineParsed(parser);
 }
 
+void MyApp::MacNewFile()
+{
+    wxDocManager::GetDocumentManager()->CreateNewDocument();
+}
+
 bool MyApp::OnInit()
 {
     if ( !wxApp::OnInit() )
@@ -229,7 +240,6 @@ bool MyApp::OnInit()
     {
         m_canvas = new MyCanvas(NULL, frame);
         m_menuEdit = CreateDrawingEditMenu();
-        docManager->CreateNewDocument();
     }
 
     CreateMenuBarForFrame(frame, menuFile, m_menuEdit);
@@ -238,6 +248,13 @@ bool MyApp::OnInit()
     frame->Centre();
     frame->Show();
 
+#if !defined( __WXMAC__ )
+    if (m_initialFiles.empty())
+        docManager->CreateNewDocument();
+#endif
+    for (size_t i = 0; i != m_initialFiles.size(); ++i)
+        docManager->CreateDocument(m_initialFiles[i], wxDOC_SILENT);
+
     return true;
 }
 
diff --git a/samples/docview/docview.h b/samples/docview/docview.h
index e598067..6b997aa 100644
--- a/samples/docview/docview.h
+++ b/samples/docview/docview.h
@@ -14,6 +14,8 @@
 
 #include "wx/docview.h"
 
+#include <vector>
+
 class MyCanvas;
 
 // Define a new application
@@ -39,6 +41,8 @@ public:
     virtual void OnInitCmdLine(wxCmdLineParser& parser) wxOVERRIDE;
     virtual bool OnCmdLineParsed(wxCmdLineParser& parser) wxOVERRIDE;
 
+    virtual void MacNewFile();
+
     // our specific methods
     Mode GetMode() const { return m_mode; }
     wxFrame *CreateChildFrame(wxView *view, bool isCanvas);
@@ -68,6 +72,8 @@ private:
     // application object itself
     void OnAbout(wxCommandEvent& event);
 
+    // a place to record filenames given on the command-line
+    std::vector<wxString> m_initialFiles;
 
     // the currently used mode
     Mode m_mode;
-- 
2.2.1

